src/stream/src/executor/aggregation/minput.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::array::stream_chunk::Ops;
+use risingwave_common::array::{ArrayImpl, Row};
+use risingwave_common::buffer::Bitmap;
+use risingwave_common::catalog::Schema;
+use risingwave_common::types::Datum;
+use risingwave_expr::expr::AggKind;
+use risingwave_storage::table::streaming_table::state_table::StateTable;
+use risingwave_storage::StateStore;
+
+use super::table_state::{
+    GenericExtremeState, ManagedArrayAggState, ManagedStringAggState, ManagedTableState,
+};
+use super::AggCall;
+use crate::common::StateTableColumnMapping;
+use crate::executor::{PkIndices, StreamExecutorResult};
+
+/// Aggregation state as a materialization of input chunks.
+///
+/// For example, in `string_agg`, several useful columns are picked from input chunks and
+/// stored in the state table when applying chunks, and the aggregation result is calculated
+/// when need to get output.
+pub struct MaterializedInputState<S: StateStore> {
+    inner: Box<dyn ManagedTableState<S>>,
+}
+
+impl<S: StateStore> MaterializedInputState<S> {
+    /// Create an instance from [`AggCall`].
+    pub fn new(
+        agg_call: &AggCall,
+        group_key: Option<&Row>,
+        pk_indices: &PkIndices,
+        col_mapping: StateTableColumnMapping,
+        row_count: usize,
+        extreme_cache_size: usize,
+        input_schema: &Schema,
+    ) -> Self {
+        Self {
+            inner: match agg_call.kind {
+                AggKind::Max | AggKind::Min | AggKind::FirstValue => {
+                    Box::new(GenericExtremeState::new(
```

Comment:
We may check whether static dispatch will help here in the future. 🥰

Reply:
I'm still refactoring these code. There's some chance that we don't need boxed dyn async trait object later, maybe.